### PR TITLE
Pandas IO Typing

### PR DESCRIPTION
**Pull request by gdj0nes**
The new code provides a typed Pandas Dataframe based on the Redshift analogue when `use_pandas=True ` with `civis.io.read_civis` based on #143. The last remaining component is to verify the testing tools. This entails updating the sql table on platform with columns with the types we would like to test.

### CODE REVIEW

File: civis/io/_tables.py [hunk 1]
```diff
@@ -33,6 +34,36 @@
...
+               'int4': np.int,
+               'int8': np.int,
+               'char': np.str,
+               'character': np.str,
+               'nchar': np.str,
+               'bpchar': np.str,
+               'varchar': np.str,
+               'character varying': np.str,
+               'nvarchar': np.str,
+               'text': np.str}
```
Comment (beckermr) [inline civis/io/_tables.py]: We might want to pick the numpy type that is appropriately sized for the redshift type. For example, for 'float8' we should prob use `np.float64`, etc. Ditto for the various int types and the rest of the float types.

File: civis/io/_tables.py [hunk 2]
```diff
@@ -33,6 +34,36 @@
...
+
+# Checkout timestamptz
```
Comment (beckermr) [inline civis/io/_tables.py]: Can you make this comment more verbose?

File: civis/io/_tables.py [hunk 3]
```diff
@@ -126,9 +160,9 @@ def read_civis(table, database, columns=None, use_pandas=False,
 
 @deprecate_param('v2.0.0', 'api_key')
 def read_civis_sql(sql, database, use_pandas=False, job_name=None,
-                   api_key=None, client=None, credential_id=None,
-                   polling_interval=None, archive=False,
-                   hidden=True, **kwargs):
+                   api_key=None, client=None, table_id=None,
```
Comment (beckermr) [inline civis/io/_tables.py]: You need to add documentation for any new parameters.

File: civis/io/_tables.py [hunk 4]
```diff
@@ -224,7 +258,12 @@ def f(x):
                                .format(script_id))
     url = outputs[0]["path"]
     if use_pandas:
-        data = pd.read_csv(url, **kwargs)
+        sortkeys, py_types, date_cols = _redshift_to_py(client, table_id)
```
Comment (beckermr) [inline civis/io/_tables.py]: You need to test here if `table_id` is `None`. If it is, then we should revert back to pandas type inference.

File: civis/io/_tables.py [hunk 5]
```diff
@@ -224,7 +258,12 @@ def f(x):
...
+        # Not sure about the best way to handle null fields
+        data = pd.read_csv(url, parse_dates=date_cols,
+                                index_col=sortkeys,
+                                converters=py_types,
+                                engine='python')
```
Comment (beckermr) [inline civis/io/_tables.py]: Can you explain this particular invocation of `pd.read_csv`? Also, we should not be setting the index of the output DF here. The convention used here IIUIC is that the DF row index is always integers and any dist or sort keys in redshift are left as normal columns. Finally, I don't think we need to revert to the python engine here, but I could be wrong.

Comment (gdj0nes) [reply civis/io/_tables.py]: I'll drop the sort keys. The python engine is much more robust and sometimes you get weird errors trying to make type conversions with the c engine, furthermore the performance difference seems negligible. When you say invocation what do you mean, the keyword args? The original was `pd.read_csv(url)`

File: civis/io/_tables.py (hunk 2 above)
Comment (gdj0nes) [reply civis/io/_tables.py]: I can just delete this comment because I got all of the aliases

File: civis/io/_tables.py (hunk 1 above)
Comment (gdj0nes) [reply civis/io/_tables.py]: I had considered that but I've found in a private repo of ours that has similar functionality and they convert all numerical to np.float but I thought it would be fine to leave the ints.

File: civis/io/_tables.py (hunk 5 above)
Comment (beckermr) [reply civis/io/_tables.py]: Yeah I just meant the overall function signature and the choice of keywords. Have you tested performance with very big tables? I think this might be where that matters but I am not sure.

File: civis/io/_tables.py (hunk 1 above)
Comment (beckermr) [reply civis/io/_tables.py]: I think here we want to preserve the type sizes properly since we don't know what users are doing with the data. We can chat about the other stuff offline. I think I understand the confusion.

**Comment by stephen-hoover** [comment]
I recommend not making any changes to `io.read_civis_sql`. That function executes arbitrary SQL, so I don't think we can have any expectation of knowing the types that come out. Stick to grabbing types for `io.read_civis`. You can use the `kwargs` in `read_civis_sql` to pass in a `dtype` parameter for `read_csv`.

**Comment by stephen-hoover** [comment]
I'm still concerned about the point that @keithing raised in #143 . Users can alter the type of columns in `read_civis`. I don't know what the best way to handle that would be. Perhaps we don't read Redshift types if `columns` is non-`None`?

**Comment by stephen-hoover** [comment]
Another comment -- in the future, please make PRs from a fork, not from a branch of this repo.

**Comment by beckermr** [comment]
Good point @stephen-hoover. RE the column transformation issue, I think the use case of asking only for some columns, but no transformations and wanting types makes a lot of sense. So I am bit worried about the guard rail as you described it.

**Comment by beckermr** [comment]
I like the idea of passing the kwargs to read_civis_sql.

**Comment by gdj0nes** [comment]
@stephen-hoover sorry for my inexperience with the conventions, I'll keep that in mind in the future.  I also like the kwargs approach.

**Comment by stephen-hoover** [comment]
Then maybe we need a switch to turn off automatic type setting? Alternately, users can already pass in `dtype` to `read_civis` if they know the types they expect. We won't be able to handle everything, and if users pass a non-`None` value to `columns`, we have no guarantees that the types in the Redshift table will be the same as the output types.

**Comment by beckermr** [comment]
I am ok with a switch. Something like `exact_types=False` or `infer_types=True`. It might be good to try and match the pandas API?

File: civis/io/_tables.py (hunk 5 above)
Comment (gdj0nes) [reply civis/io/_tables.py]: I have a table that's over 1.5GB and I can't say I noticed any difference switching between the two but it did crash when there was an integer column that had some funny formatting when using the C engine. See preformance comparison: [here](https://stackoverflow.com/questions/29678154/why-is-numpy-pandas-parsing-of-a-csv-file-with-long-lines-so-slow)

**Comment by gdj0nes** [comment]
@stephen-hoover if we want users to pass the dtypes manually should we create a helper function that grabs the column dtypes for them outside of `read_civis` to expedite the process?

@beckermr if we use your approach would we add this keyword to the `read_civis` function and then if it is `True` grab the table meta in `read_civis` and then pass it as kwargs to `read_civis_sql`?

Is there a consensus on how to go forward with this?

**Comment by stephen-hoover** [comment]
I don't think it's necessary to add a public helper function here.

What about the following: Add an explicit `dtype` parameter to `read_civis`. Default it to "redshift". If the value is "redshift", then replace it with data types read from the specified table. Then pass `dtype` (possibly replaced with types read from Redshift, possibly not) to `read_civis_sql`.

**Comment by beckermr** [comment]
I like this suggestion @stephen-hoover. Do we want to default to None?

**Comment by keithing** [comment]
Would it be a breaking change if we default a new `dtype` parameter to `redshift`?

**Comment by stephen-hoover** [comment]
It might change the function's output, and it could break previously functioning code if users were using `columns` to modify output types. Maybe it should default to `None`. Reading types from Redshift seems like a nicer default, though -- perhaps we can make a note to switch the default to "redshift" if/when we make a v2 release?

**Comment by gdj0nes** [comment]
I'm worried that if we have the default being redshift and a user is currently passing their own dtypes through kwargs then we overwrite their dtypes without them knowing. It also seems that dtypes should boolean because it has no other use case then grabbing redshift types.

**Comment by stephen-hoover** [comment]
If users are already setting `dtype`, that will override our default. My suggestion is that we add an extra option for what users could pass to `dtype`, not create a new parameter. (See https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html for documentation on the `dtype` parameter.)

**Comment by beckermr** [comment]
Yep @stephen-hoover is right. Here is what we would have

- `dtype='redshift'` - infer types
- `dtype=None` - let pandas infer the types as usual (default)
- `dtype=<anything else accepted by pandas>` - let pandas do its thing (item above is a special case of this)

**Comment by gdj0nes** [comment]
@beckermr If it is anything else excepted by pandas should we change it to `pd_options` or some other more descriptive name?

**Comment by beckermr** [comment]
Nah. Pandads calls it dtype, so let's call it dtype. It is anything else accepted by the pandas dtype pasrameter on read_csv, so this makes sense. Sorry I was unclear about that!

**Comment by gdj0nes** [comment]
Any thoughts on how to update the testing suite?

**Comment by beckermr** [comment]
We should do this in person. Feel free to add 30 min to my calendar.

File: civis/io/_tables.py [hunk 6]
```diff
@@ -115,9 +149,24 @@ def read_civis(table, database, columns=None, use_pandas=False,
...
+        if dtype == 'redshift':
+            table_id = client.get_table_id(table=table, database=database)
+            py_types, date_cols = _redshift_to_py(client, table_id)
+            pd_kwargs = {'parse_dates': date_cols,
+                         'dtype': py_types,
+                         'engine': 'python',
+                         'infer_datetime_format': True}
+            kwargs = kwargs.update(pd_kwargs)
+        elif type(dtype) == dict:
+            kwargs = kwargs.update(pd_kwargs)
```
Comment (beckermr) [inline civis/io/_tables.py]: This line is wrong. You want `kwargs['dtype'] = dtype` I think.

File: civis/io/_tables.py [hunk 7]
```diff
@@ -115,9 +149,24 @@ def read_civis(table, database, columns=None, use_pandas=False,
...
+        else:
```
Comment (beckermr) [inline civis/io/_tables.py]: No need for this block.

File: civis/io/_tables.py [hunk 8]
```diff
@@ -115,9 +149,24 @@ def read_civis(table, database, columns=None, use_pandas=False,
...
+        elif type(dtype) == dict:
+            kwargs = kwargs.update(pd_kwargs)
+        else: 
+            continue
+
     sql = _get_sql_select(table, columns)
     data = read_civis_sql(sql=sql, database=database, use_pandas=use_pandas,
-                          job_name=job_name, client=client,
+                          job_name=job_name, client=client, 
+                          table_id=table_id,
```
Comment (beckermr) [inline civis/io/_tables.py]: This can be removed.

File: civis/io/_tables.py [hunk 9]
```diff
@@ -127,7 +176,7 @@ def read_civis(table, database, columns=None, use_pandas=False,
 @deprecate_param('v2.0.0', 'api_key')
 def read_civis_sql(sql, database, use_pandas=False, job_name=None,
                    api_key=None, client=None, credential_id=None,
-                   polling_interval=None, archive=False,
+                   polling_interval=None, archive=False,
```
Comment (beckermr) [inline civis/io/_tables.py]: flake8 will complain about this whitespace change...ugh :(

**Comment by gdj0nes** [comment]
I've encountered an issue related to null types and integers. In the case that there is a`null` type in an integer column, the converter will fail because `NaN`s are floats in Numpy. Possible Solutions:

1. Try block reverts back to the default behavior with a warning message
2. Try block and switch to all (we don't know which column caused the problem)  ints to floats

**Comment by stephen-hoover** [comment]
Ugh, that's annoying. Another option is to read the DataFrame with all integer types set to float and then try one-by-one to cast the columns back to integers. It seems like that might be faster than a try/except on reading the entire DataFrame. Still potentially slow, though.

We could even consider reading all integer types as floats and not bothering to keep them as ints.

Note that with anything involving reading integer types as floats, we need to use more bits of precision in the float types than in the Redshift integer type. Otherwise we'll end up losing precision. Some of the float's bits are used for the exponent, so it can't exactly represent all of the integers in an integer type with the same number of bits.

**Comment by beckermr** [comment]
Yeah I actually like Stephen's suggestion. It is also probably the only workable one.

**Comment by beckermr** [comment]
The one I meant was just reading int types as float and then casting back if we can.

**Comment by gdj0nes** [comment]
Thoughts on how to modify the testing function? Should I just update `scratch.api_client_test_fixture` with a var char, bool, int, float, null int, and date?

**Comment by jseabold** [comment]
Hey, cool! Before you get too far, I'd recommend at least having a look at how sqlalchemy handles this. You'll get all the standard types from postgres + any redshift specific ones via redshift-sqlalchemy. Here's some helper functions to make a pandas-ready dtype from redshift. Should cover every type, but I'm not positive I didn't stop along the way.

https://github.com/civisanalytics/modeling-dsrd/blob/dask-redshift/dask-redshift/dask_redshift/dask_redshift.py#L12

As for integers and nulls, the best you can do is [check if an integer is nullable](https://github.com/civisanalytics/modeling-dsrd/blob/dask-redshift/dask-redshift/dask_redshift/dask_redshift.py#L69) and use a float. Otherwise, we'd have to pass over all of the data.

**Comment by jseabold** [comment]
I remember at least one place I stopped. The char types can have an encoding associated with them, I didn't try to use it yet.

**Comment by stephen-hoover** [comment]
What exactly is the "notnull" column in pg_table_def? The docs say "True if the column has a NOT NULL constraint.", but what is a "NOT NULL" constraint? I'm a bit concerned by that conversion. Forcing ints into a float32 will make some distinct integers no longer distinct. This has caused bugs for us in the past.

**Comment by jseabold** [comment]
>  I'm a bit concerned by that conversion. Forcing ints into a float32 will make some distinct integers no longer distinct. This has caused bugs for us in the past.

Ah, right.

From the docs:

>>> NOT NULL specifies that the column is not allowed to contain null values. NULL, the default, specifies that the column accepts null values. IDENTITY columns are declared NOT NULL by default.

**Comment by beckermr** [comment]
Generally the client doesn’t uses direct database connection since it goes through the API. Is it possible to use sqlalchemy without a database connection? Or do you just mean look at the code for how they do this?

**Comment by stephen-hoover** [comment]
I've checked some of my own tables, and I have some with integer ID columns which are marked nullable (but which don't have nulls). We should make sure to represent integer columns with enough bits of precision that we can represent all allowable integers.

File: civis/io/_tables.py [hunk 10]
```diff
@@ -727,3 +762,48 @@ def f(x):
...
+        'int4': np.int32,
+        'int8': np.int64,
+        'char': np.str,
+        'character': np.str,
+        'nchar': np.str,
+        'bpchar': np.str,
+        'varchar': np.str,
+        'character varying': np.str,
+        'nvarchar': np.str,
+        'text': np.str}
```
Comment (beckermr) [inline civis/io/_tables.py]: @jseabold I think the type map here is pretty complete already. We just have to deal with nulls. Probs best to read as float64 and then do a quick check and convert back to the right in type if there are not any nulls.

Comment (beckermr) [reply civis/io/_tables.py]: redshift by default uses UTF-8 encoding. Thus I think we should make sure to account for that and then punt on the rest.

Comment (gdj0nes) [reply civis/io/_tables.py]: I've added a utf-8 encoding parameter to the kwargs

**Comment by gdj0nes** [comment]
It turns out that you can't cast back to integer after the import because the problem still remains that NaNs can't be converted to Int types.

**Comment by beckermr** [comment]
You need to test for nans first and if there are none, then cast back, otherwise don’t.

**Comment by gdj0nes** [comment]
Are there any thoughts about how the test should be incorporated?

File: civis/io/_tables.py [hunk 11]
```diff
@@ -727,3 +765,48 @@ def f(x):
 
         fut.add_done_callback(f)
     return fut
+
+
+def _redshift_to_py(civ_client, tableid):
```
Comment (stephen-hoover) [inline civis/io/_tables.py]: This library has the convention of naming `APIClient` objects as `client`. You should use that instead of `civ_client` here.

File: civis/io/_tables.py [hunk 12]
```diff
@@ -115,6 +127,23 @@ def read_civis(table, database, columns=None, use_pandas=False,
     if client is None:
         # Instantiate client here in case users provide a (deprecated) api_key
         client = APIClient(api_key=api_key, resources='all')
+
+    if use_pandas == True:
+        if dtype == 'redshift':
+            if NO_NUMPY:
+                raise ImportError("`dtype='redshift'` specified but numpy not installed.")
```
Comment (stephen-hoover) [inline civis/io/_tables.py]: Is this `numpy` check needed? If we have `pandas`, then we know that we have `numpy`.

File: civis/io/_tables.py [hunk 13]
```diff
@@ -727,3 +765,48 @@ def f(x):
 
         fut.add_done_callback(f)
     return fut
+
+
+def _redshift_to_py(civ_client, tableid):
+
+    SQL_PANDAS_MAP = {
```
Comment (stephen-hoover) [inline civis/io/_tables.py]: I suggest moving this dictionary and `DATE_TYPES` into module-level constants. They seem like something worth putting at the top of the file.

File: civis/tests/test_io.py [hunk 14]
```diff
@@ -152,6 +159,61 @@ def test_read_civis_pandas(self, *mocks):
...
+
+        # Construct all of these column types
+
+        pd_data = {}
+        for num, (name, type) in enumerate(SQL_PANDAS_MAP.items()):
+            if name in ['boolean', 'bool']:
+                value = [True]
+            elif name in ['char', 'character', 'nchar', 'bpchar', 'varchar',
+                          'character varying', 'nvarchar', 'text']:
+                value = ['foo']
```
Comment (stephen-hoover) [inline civis/tests/test_io.py]: Try making this something like `"13"`. If pandas can infer types, then we aren't testing all of the new code.

File: civis/tests/test_io.py [hunk 15]
```diff
@@ -152,6 +159,61 @@ def test_read_civis_pandas(self, *mocks):
...
+            else:
+                value = [num]
+            pd_data[name] = value
+
+        expected = pd.DataFrame
+
+        df = civis.io.read_civis('scratch.api_client_test_fixture',
+                                 'redshift-general', use_pandas=True,
+                                 dtype='redshift', polling_interval=POLL_INTERVAL)
+        assert df.equals(expected)
```
Comment (stephen-hoover) [inline civis/tests/test_io.py]: Doing this test in the same way as `test_read_civis_pandas` would require recording a new VCR cassette.

File: civis/io/_tables.py [hunk 16]
```diff
@@ -115,6 +127,23 @@ def read_civis(table, database, columns=None, use_pandas=False,
...
+    if use_pandas == True:
+        if dtype == 'redshift':
+            if NO_NUMPY:
+                raise ImportError("`dtype='redshift'` specified but numpy not installed.")
+            else:
+                table_id = client.get_table_id(table=table, database=database)
+                py_types, date_cols = _redshift_to_py(client, table_id)
+                pd_kwargs = {'parse_dates': date_cols,
+                             'dtype': py_types,
+                             'engine': 'python',
```
[the remark this answers is not in the document]
Comment (gdj0nes) [reply civis/io/_tables.py]: I defaulted to this because it is more robust and the performance benefits of the C engine seem negligible for our data

File: civis/io/_tables.py [hunk 17]
```diff
@@ -115,6 +127,23 @@ def read_civis(table, database, columns=None, use_pandas=False,
...
+                             'infer_datetime_format': True,
+                             'encoding': 'utf8'}
```
[the remark this answers is not in the document]
Comment (gdj0nes) [reply civis/io/_tables.py]: see previous comment

File: civis/io/_tables.py [hunk 18]
```diff
@@ -115,6 +127,23 @@ def read_civis(table, database, columns=None, use_pandas=False,
...
+                kwargs.update(pd_kwargs)
+        elif type(dtype) == dict:
+            kwargs['dtype'] = dtype
```
[the remark this answers is not in the document]
Comment (gdj0nes) [reply civis/io/_tables.py]: That equality check is good to know

I'm not sure what you mean by allow string types

[the remark this answers is not in the document]
Comment (stephen-hoover) [reply civis/io/_tables.py]: I mean that valid inputs include both `float` and `"float"`.

File: civis/io/_tables.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (stephen-hoover) [reply civis/io/_tables.py]: I'm not comfortable altering this default, especially since it would only apply to a special subset of circumstances and it's hidden from users. Is there a specific problem you're solving with this? Users have the option to modify the "engine" parameter themselves if they so choose.

File: civis/io/_tables.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (stephen-hoover) [reply civis/io/_tables.py]: You mean "because it is more robust"? Can you give an example of a problem that this solves?

File: civis/io/_tables.py [hunk 19]
```diff
@@ -224,7 +252,17 @@ def f(x):
                                .format(script_id))
     url = outputs[0]["path"]
     if use_pandas:
+        user_dtypes = kwargs.pop('dtype')
+        # Only for types for str and bool
+        kwargs['dtype'] = {k:v for k,v in user_dtypes.items()
+                           if v in [np.str, np.bool]}
```
[the remark this answers is not in the document]
Comment (gdj0nes) [reply civis/io/_tables.py]: In addition to the subclass should it search for "int" provided as a string?

[the remark this answers is not in the document]
Comment (stephen-hoover) [reply civis/io/_tables.py]: No, `np.issubdtype` will handle that for you!

File: civis/io/_tables.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (gdj0nes) [reply civis/io/_tables.py]: I remember having to change to it but I can't remember why I had to make this change. This is not the first time I have encountered this problem either. If we encounter the error that is raised when the C engine fails it would probably be nice to raise a flag that the user try the python engine

File: civis/io/_tables.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (gdj0nes) [reply civis/io/_tables.py]: Learning so much here

File: civis/io/_tables.py [hunk 20]
```diff
@@ -224,7 +252,17 @@ def f(x):
...
         data = pd.read_csv(url, **kwargs)
+        dtype_cols = [c for c in list(data) if c in set(user_dtypes.keys())]
+        for col in dtype_cols:
+            are_nulls = data[col].isnull().any()
+            if not are_nulls:
+                data[col] = data[col].astype(user_dtypes[col])
```
[the remark this answers is not in the document]
Comment (gdj0nes) [reply civis/io/_tables.py]: Good point, I'll drop it

File: civis/io/_tables.py [hunk 21]
```diff
@@ -32,11 +39,41 @@
...
+    'bigint': np.int64,
+    'decimal': np.float64,
+    'numeric': np.float64,
+    'float': np.float64,
+    'float4': np.float32,
+    'float8': np.float64,
+    'real': np.float32,
+    'double precision': np.float64,
+    'boolean': np.bool,
+    'bool': np.bool,
```
Comment (beckermr) [inline civis/io/_tables.py]: Doesn't redshift output 't' and 'f' for boolean values? http://docs.aws.amazon.com/redshift/latest/dg/r_Boolean_type.html

I think that means this would fail. I thought @stephen-hoover ran into this in CivisML?

Comment (beckermr) [reply civis/io/_tables.py]: Yeah this won't work.

```python
In [1]: np.bool('t')
Out[1]: True

In [2]: np.bool('f')
Out[2]: True
```

**Comment by beckermr** [comment]
@gdj0nes @stephen-hoover Can we get this into the 2.0 release?

I talked with @gdj0nes about this today and we agreed that the 2.0 release is a good time since it might be a big change.

**Comment by beckermr** [comment]
Right. My thought was that if people call `read_civi` with no column arguments, then we infer the types for that case. Anything is too hard I agree. This might too weird of a special case to have wrapped up in the client I agree. 

We could add a function to return table types that are numpy compatible. I now cargo cult this around everywhere.

**Comment by stephen-hoover** [comment]
Scanning through this PR again -- I like the approach of forcing users to opt-in to this with an explicit call to the `dtypes` parameter. I think there is a version of this PR which would work. I'd have to read through in more detail to say how close this is. If someone were to take it up again, it looks like there's still an outstanding issue with handling booleans, and it needs to be rebased on the current master.

File: civis/io/_tables.py [hunk 22]
```diff
@@ -115,6 +156,20 @@ def read_civis(table, database, columns=None, use_pandas=False,
...
+            table_id = client.get_table_id(table=table, database=database)
+            py_types, date_cols = _redshift_to_py(client, table_id)
+            pd_kwargs = {'parse_dates': date_cols,
+                         'dtype': py_types,
+                         'engine': 'python',
+                         'infer_datetime_format': True,
+                         'encoding': 'utf8'}
+            kwargs.update(pd_kwargs)
+        else:
+            kwargs['dtype'] = dtype
```
Comment (keithing) [inline civis/io/_tables.py]: Will this work if the user passes `columns`?  I think it will at least break if the users passes sql into the columns.

Comment (beckermr) [reply civis/io/_tables.py]: It would not work with columns in general no.

**Comment by beckermr** [comment]
I’m ok with optin and only on read_civis. I say we just let people do what they want wrt columns.

**Comment by beckermr** [comment]
Alright. I am looking at this and thinking it is not worth it. We could add a function to the client to extract table types but given how we do exports this is not going to work. 

If we start to use parquet or some other typed format on the backend, this would solve this issue and speed up the export etc. 

I vote we close this.

**Comment by beckermr** [comment]
Alright. Closing. Feel free to reopen if you feel differently OFC!
